added feature "discount-codes"

simarsingh24 opened the pull request:
Mentions #2296 
Screenshots for the change: 
![screenshot_20180204-030739](https://user-images.githubusercontent.com/16715499/35771842-1b8cfcf0-0959-11e8-8237-fecfc23c1058.png)

simarsingh24 commented:
@iamareebjamal @srv-twry @mayank8318 @nikit19  Is this a good way to show discount-codes also this is initial steps (the indentation and refactors will be done later, UI will be improved etc )
This is just for review whether I am working in the right direction, the core logic for getting the discount-codes is right?

simarsingh24 commented:
@iamareebjamal would that be okay if discount codes are shown here, if yes I would proceed further finalizing everything  (UI will be changed a little and some more details of discount codes will be added)

iamareebjamal commented:
I think AboutFragment has lost its actual use case, don't add it there. Create a new section

simarsingh24 commented:
![screenshot_20180206-224915](https://user-images.githubusercontent.com/16715499/35873638-4f2b460a-0b90-11e8-82ae-70f1c1a0ee68.png)
This is how it looks now, created a new section in navigation for discount codes.
@iamareebjamal please tell me out of these which fields of discount codes should be displayed here since the API is mostly returning null or 0 values.

iamareebjamal commented:
Show the fields which only have non null values

dr0pdb reviewed (changes requested):
Please make the download lazy i.e. Download should be done after the user enters the fragment.
Also remove the dependency on the event bus pattern.

simarsingh24 commented:
@srv-twry yeah that i will do also because in order to view dicount codes Authentication is necessary ie user should be logged in, so that check and downloading part will be moved to fragment

simarsingh24 commented:
screen shots after changes : 
![screenshot_20180210-113300](https://user-images.githubusercontent.com/16715499/36059255-a3bc9102-0e5a-11e8-8c18-65cf7d58dcdd.png)

simarsingh24 commented:
squashed commits and resolved conflicts

ghost commented:
Hi @simarsingh24!

It looks like one or more of your builds have failed.
I've added the relevant info below to save you some time.

- [Build 5110.1](https://travis-ci.org/fossasia/open-event-android/jobs/339770754) has `failed`

simarsingh24 commented:
@iamareebjamal please review

iamareebjamal commented:
1. Don't use colors so much, it decreases the themability of the app
2. The code of discount coupon is taking up a lot of space, reduce it
3. Ticket No. has no significance to the user. Actually unless the tickets are loaded and linked in the app, a user can just "see" this information, so we need to hide it from user and only make it available to dev version if discount url is missing

simarsingh24 commented:
changes made:
1) UI cleaned (simple white background CardView)
2) Removed TicketNo from displaying
@iamareebjamal any ideas how can i reduce the code space ?

iamareebjamal commented:
I was talking about the discount code width in UI

simarsingh24 commented:
@iamareebjamal  haha sorry I misunderstood 😆

simarsingh24 commented:
screen shots after the change:
![screenshot_20180210-153907](https://user-images.githubusercontent.com/16715499/36060928-b7c24246-0e78-11e8-8ded-aaa144e7ca13.png)

dr0pdb reviewed (changes requested):
I would suggest you please look at the recently merged implementation of the NotificationsFragment for reference.
You have missed certain points IMO.

1. When i asked you to make it lazy, i also meant don't use the EventBus. You can make a repository to handle these.

2. You haven't implemented the swipe to refresh.

3. If the user isn't logged in , then kindly redirect him/her to the login activity. This would make it even better.

PS: Pardon if you are still working on it and planned on implementing these.

simarsingh24 commented:
@srv-twry made required changes!

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/common/api/DataDownloadManager.java:
```diff
@@ -51,4 +51,5 @@ public void downloadMicrolocations() {
     public void downloadSessionTypes() {
         APIClient.getOpenEventAPI().getSessionTypes().enqueue(new SessionTypeListResponseProcessor());
     }
+
```
You can delete this line. Redundant change.

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/common/api/DownloadCompleteHandler.java:
```diff
@@ -211,6 +211,7 @@ public void onMicrolocationsDownloadDone(MicrolocationDownloadEvent event) {
         public void onSessionTypesDownloadDone(SessionTypesDownloadEvent event) {
             onDownloadDone(event, R.string.session_types);
         }
+
```
Same here.

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/common/events/DiscountCodesDownloadEvent.java:
```diff
@@ -0,0 +1,8 @@
+package org.fossasia.openevent.common.events;
+
+public class DiscountCodesDownloadEvent extends DownloadEvent {
+
```
We don't need this class anymore. Right ?

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,160 @@
+
+        final View view = super.onCreateView(inflater, container, savedInstanceState);
+        setUpRecyclerView();
+        //set up view model
```
This comment should be below. Before line 60.

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,160 @@
+                downloadDiscountCodes();
+            }
+            loadData();
+
```
I think this line isn't required. Also give proper spaces on line 61.

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,160 @@
+
+    private void downloadDiscountCodes() {
+        discountFragmentViewModel.downloadDiscountCodes().observe(this, this::onDiscountCodeDownloadDone);
+
```
remove this line.

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    private void handleVisibility() {
+
```
no need for this line.

dr0pdb commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,160 @@
+            this.discountCodes.addAll(discountCodes);
+            discountCodesListAdapter.notifyDataSetChanged();
+            handleVisibility();
+            Timber.d("Total discount codes +"+this.discountCodes.size());
```
Please do this logging inside the ViewModel if possible.

dr0pdb reviewed (changes requested):
Lot of formatting errors. Please find others also and fix them :+1:

simarsingh24 commented:
@srv-twry  oh i removed these but forgot to stage the changes, will do it right away!

simarsingh24 commented:
Should I change the navigation icon for discount code too? Right now its set to "Home" icon

dr0pdb commented:
Yes, please.

simarsingh24 commented:
@iamareebjamal please review

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,155 @@
+        Toast.makeText(getContext(), "User need to be logged in!", Toast.LENGTH_SHORT).show();
+        Intent intent = new Intent(getActivity(), LoginActivity.class);
+        startActivity(intent);
+    }
```
Several formatting related issues, please resolve

simarsingh24 replied:
made some formatting changes, but couldn't find anything in this highlighted area

simarsingh24 replied:
@iamareebjamal  could you please confirm?

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,152 @@
+        setUpRecyclerView();
+        Utils.registerIfUrlValid(swipeRefreshLayout, this, this::refresh);
+        discountFragmentViewModel = ViewModelProviders.of(this).get(DiscountFragmentViewModel.class);
+        if(AuthUtil.isUserLoggedIn()) {
```
here

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,152 @@
+        Utils.registerIfUrlValid(swipeRefreshLayout, this, this::refresh);
+        discountFragmentViewModel = ViewModelProviders.of(this).get(DiscountFragmentViewModel.class);
+        if(AuthUtil.isUserLoggedIn()) {
+            if(NetworkUtils.haveNetworkConnection(getContext())) {
```
here

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountCodeFragment.java:
```diff
@@ -0,0 +1,152 @@
+                downloadDiscountCodes();
+            }
+            loadData();
+        }else {
```
here

simarsingh24 commented:
made requested changes and squashed commits

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,83 @@
+        if (!TextUtils.isEmpty(field.trim())) {
+            textView.setVisibility(View.VISIBLE);
+            field = prefix + field;
+            textView.setText(field);
```
Directly set prefix + field in TextView

simarsingh24 replied:
done

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+
+    private void setStringField(TextView textView, String field, String prefix) {
```
Rename the function appropriately

simarsingh24 replied:
changed

simarsingh24 commented:
squashed commits

iamareebjamal commented on android/app/src/main/res/values/dimens.xml:
```diff
@@ -44,7 +44,7 @@
     <dimen name="text_size_expanded_title">22sp</dimen>
     <dimen name="text_size_extra_large">18sp</dimen>
-    <dimen name="text_size_large">16sp</dimen>
+    <dimen name="text_size_large">20sp</dimen>
```
Why is large 20sp and extra large 18 sp?

simarsingh24 replied:
reverting back to 16sp

simarsingh24 committed:
added discount codes


changed nav icon


indentation changes


indentation changes


indentation changes


refactor changes


minor change

simarsingh24 commented:
@iamareebjamal  please review

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,82 @@
+        ButterKnife.bind(this, itemView);
+
+
+    }
```
extra lines

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,82 @@
+
+    public void bindDiscountCode(DiscountCode discountCode) {
+
+        String code = Utils.checkStringEmpty(discountCode.getCode());
```
extra lines

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,82 @@
+        setStringFieldWithPrefix(discountUsedFor, usedFor, "Used For : ");
+        setStringFieldWithPrefix(this.discountUrl, discountUrl, "Discount URL : ");
+
+    }
```
Extra line

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,82 @@
+        setStringFieldWithPrefix(discountCodeTextView, code, "");
+        setStringFieldWithPrefix(discountCodeValue, "Value : "+value, "");
+        setStringFieldWithPrefix(discountMinQuantity, "Minimum Quantity : "+minQuantity, "");
+        setStringFieldWithPrefix(discountMaxQuantity, "Maximum Quantity :"+maxQuantity, "");
```
Spacing between +

I'd recommend using template strings

simarsingh24 replied:
do you mean use of strings.xml
like this :
`<string name="MinQuanString">Minimum Quantity : %1$d </string>`

iamareebjamal replied:
Yes

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/core/discount/DiscountViewHolder.java:
```diff
@@ -0,0 +1,82 @@
+        setStringFieldWithPrefix(discountCodeTextView, code, "");
+        setStringFieldWithPrefix(discountCodeValue, "Value : "+value, "");
+        setStringFieldWithPrefix(discountMinQuantity, "Minimum Quantity : "+minQuantity, "");
+        setStringFieldWithPrefix(discountMaxQuantity, "Maximum Quantity :"+maxQuantity, "");
```
Rather than using `""` empty strings, pass null and check for the same

simarsingh24 committed:
indentation changes

simarsingh24 committed:
used string template

simarsingh24 commented:
@iamareebjamal discount codes are not being displayed now! can you point out what went wrong in the last commit ?
